### PR TITLE
Climate PID Autotune Logging fixes

### LINKED ISSUE #3858: Multiple `pid` climates cannot be differentiated in the logs
### The problem

I made a central heating system with 7 climates, each with its own PID. When I try to run Autotune for all of them, unfortunately the log doesn't show which message belongs to which climate:
```
[09:07:09][I][pid.autotune:134]: PID Autotune:
[09:07:09][I][pid.autotune:179]:   Autotune is still running!
[09:07:17][I][pid.autotune:134]: PID Autotune:
[09:07:17][I][pid.autotune:179]:   Autotune is still running!
[09:07:19][I][pid.autotune:134]: PID Autotune:
[09:07:19][I][pid.autotune:179]:   Autotune is still running!
[09:07:27][I][pid.autotune:134]: PID Autotune:
[09:07:27][I][pid.autotune:179]:   Autotune is still running!
[09:07:29][I][pid.autotune:134]: PID Autotune:
[09:07:29][I][pid.autotune:179]:   Autotune is still running!
[09:07:37][I][pid.autotune:134]: PID Autotune:
[09:07:37][I][pid.autotune:179]:   Autotune is still running!
[09:07:39][I][pid.autotune:134]: PID Autotune:
[09:07:39][I][pid.autotune:179]:   Autotune is still running!
[09:07:47][I][pid.autotune:134]: PID Autotune:
[09:07:47][I][pid.autotune:179]:   Autotune is still running!
[09:07:49][I][pid.autotune:134]: PID Autotune:
[09:07:49][I][pid.autotune:179]:   Autotune is still running!
[09:07:57][I][pid.autotune:134]: PID Autotune:
[09:07:57][I][pid.autotune:179]:   Autotune is still running!
[09:07:59][I][pid.autotune:134]: PID Autotune:
[09:07:59][I][pid.autotune:179]:   Autotune is still running!
[09:08:07][I][pid.autotune:134]: PID Autotune:
[09:08:07][I][pid.autotune:179]:   Autotune is still running!
...
```
Although each climate has it's own `id` and `name`, it can't be told which of these logger messages belong to which climate...

It's not really useful to run the autotunes separately because they can influence each other (rooms of the same building) so for accurate measurement it's required to run them all in parallel.

Another annoying issue is that the INFO message "_Autotune is still running!_" is shown every 10 seconds per instance. An autotune procedure for a heating system takes several hours. Log is flooded even in INFO level.

With multiple PIDs running Autotune it gets even worse, and when one completes, the printed results disappear due to overscrolling as the other PID instances still running keep flooding the log with the same still running message.

After hours of running, result appears like this:

```
 PID Autotune finished!
 PID Autotune:
   State: Succeeded!
   Could not reliable determine oscillation amplitude, PID parameters may be inaccurate!
     Please make sure you eliminate all outside influences on the measured temperature.
   Oscillation Frequency is not symmetrical. PID parameters may be inaccurate!
     This is usually because the heat and cool processes do not change the temperature at the same rate.
     Please try reducing the positive_output value (or increase negative_output in case of a cooler)
   Calculated PID parameters ("Ziegler-Nichols PID" rule):

   control_parameters:
     kp: 0.59220
     ki: 0.00012
     kd: 735.99213
```
It can't be told these results belong to which PID instance.

Control parameters are used immediately after finish, so they can be noted down if one prepares sensors for these in advance even if the message is lost, however the very useful textual information can't be identified which PID it belongs to, and is lost when logs scroll over.

**Suggestion**

1. move the "_Autotune is still running!_" message to a Debug/Verbose level, instead of INFO, to avoid spamming the log
2. display the message "_For PID autotuner you need to set AUTO (also called heat/cool) mode!_" only once (at the start of Autotune), not repeatedly, to avoid spamming the log
3. add a tag or a string with _Name_ or _ID_ to the logs so we know from which PID climate instance the log messages are coming from

This way, we could start Autotune for multiple instances and with the log level set to INFO, we could have the separately identifiable results after several hours without them being scrolled over.

### Which version of ESPHome has the issue?

2022.11.3

### What type of installation are you using?

Home Assistant Add-on

### Which version of Home Assistant has the issue?

2022.11

### What platform are you using?

ESP32

### Board

WirelessTag WT32-ETH01

### Component causing the issue

pid

### Example YAML snippet

```yaml
climate:
- platform: pid
  id: heat_01_pid
  name: ${zone_1_name} fűtés PID
  visual:
    min_temperature: 14
    max_temperature: 25
    temperature_step: 0.1
  sensor: heat_01_temp
  default_target_temperature: ${default_target_temp}
  heat_output: heat_01_pwm
  control_parameters:
    kp: 0.0
    ki: 0.0
    kd: 0.0

- platform: pid
  id: heat_02_pid
  name: ${zone_2_name} fűtés PID
  visual:
    min_temperature: 14
    max_temperature: 25
    temperature_step: 0.1
  sensor: heat_02_temp
  default_target_temperature: ${default_target_temp}
  heat_output: heat_02_pwm
  control_parameters:
    kp: 0.0
    ki: 0.0
    kd: 0.0

- platform: pid
  id: heat_03_pid
  name: ${zone_3_name} fűtés PID
  visual:
    min_temperature: 14
    max_temperature: 25
    temperature_step: 0.1
  sensor: heat_03_temp
  default_target_temperature: ${default_target_temp}
  heat_output: heat_03_pwm
  control_parameters:
    kp: 0.0
    ki: 0.0
    kd: 0.0

- platform: pid
  id: heat_04_pid
  name: ${zone_4_name} fűtés PID
  visual:
    min_temperature: 14
    max_temperature: 25
    temperature_step: 0.1
  sensor: heat_04_temp
  default_target_temperature: ${default_target_temp}
  heat_output: heat_04_pwm
  control_parameters:
    kp: 0.0
    ki: 0.0
    kd: 0.0

- platform: pid
  id: heat_05_pid
  name: ${zone_5_name} fűtés PID
  visual:
    min_temperature: 14
    max_temperature: 25
    temperature_step: 0.1
  sensor: heat_05_temp
  default_target_temperature: ${default_target_temp}
  heat_output: heat_05_pwm
  control_parameters:
    kp: 0.0
    ki: 0.0
    kd: 0.0

- platform: pid
  id: heat_06_pid
  name: ${zone_6_name} fűtés PID
  visual:
    min_temperature: 14
    max_temperature: 25
    temperature_step: 0.1
  sensor: heat_06_temp
  default_target_temperature: ${default_target_temp}
  heat_output: heat_06_pwm
  control_parameters:
    kp: 0.0
    ki: 0.0
    kd: 0.0

- platform: pid
  id: heat_07_pid
  name: ${zone_7_name} fűtés PID
  visual:
    min_temperature: 14
    max_temperature: 25
    temperature_step: 0.1
  sensor: heat_07_temp
  default_target_temperature: ${default_target_temp}
  heat_output: heat_07_pwm
  control_parameters:
    kp: 0.0
    ki: 0.0
    kd: 0.0
```

### Anything in the logs that might be useful for us?

_logs pasted in the section above_

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by patrickcollins12**
# What does this implement/fix?

[User requested](https://github.com/esphome/issues/issues/3858) changes to PID Autotune logging, to allow it to work with multiple concurrent PID Climate devices. 

## Types of changes

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes [this issue](https://github.com/esphome/issues/issues/3858)

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** no documentation changes needed.

## Test Environment

- [x] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040

## Example entry for `config.yaml`:

```yaml
#### FOR DEBUG 
## This is standard functionality, nothing added.

output:
  - platform: ledc
    id: console_heat_speed
    pin: GPIO32
  - platform: ledc
    id: console_fan_speed
    pin: GPIO25
climate:
  - platform: pid
    name: "Console Fan Thermostat"
    id: console_thermostat
    sensor: console_fan_temperature
    default_target_temperature: 21°C
    heat_output: console_heat_speed
    cool_output: console_fan_speed

switch:
  - platform: template
    name: "PID Climate Autotune"
    turn_on_action:
      - climate.pid.autotune: console_thermostat


```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [ ] N/A

### CODE REVIEW

**Comment by nagyrobi** [comment]
@patrickcollins12 great!

Note:
- _!!! For PID autotuner you need to set AUTO (also called heat/cool) mode!_ message is missing the `%s:", name_.c_str()`
- Maybe an _Autotune has started. This can take a long time._ kind of message would be good to have at he very beginning once (also with name), to have _one_ feedback that things have started.

**Comment by nagyrobi** [comment]
PID config:

```yaml
climate:
- platform: pid
  id: heat_01_pid
  name: Iroda külső fűtés PID
  visual:
    min_temperature: 14
    max_temperature: 25
    temperature_step: 0.1
  sensor: heat_01_temp
  default_target_temperature: 19
  heat_output: heat_01_pwm
  control_parameters:
    kp: 0.0
    ki: 0.0
    kd: 0.0
```

Started Autotune for it, took 6 hours:

```
[19:58:27][W][pid.climate:158]: !!! For PID autotuner you need to set AUTO (also called heat/cool) mode!
[20:02:32][I][ota:113]: Boot seems successful, resetting boot loop counter.
[02:07:09][I][pid.autotune:119]: PID Autotune finished!
[02:07:09][I][pid.autotune:135]: PID Autotune - Iroda külső fűtés PID (iroda_kls_fts_pid):
[02:07:09][I][pid.autotune:136]:   State: Succeeded!
[02:07:09][W][pid.autotune:144]:   Oscillation Frequency is not symmetrical. PID parameters may be inaccurate!
[02:07:09][W][pid.autotune:147]:     This is usually because the heat and cool processes do not change the temperature at the
[02:07:09][W][pid.autotune:149]:     Please try reducing the positive_output value (or increase negative_output in case of a
[02:07:09][I][pid.autotune:157]:   Calculated PID parameters ("Ziegler-Nichols PID" rule):
[02:07:09][I][pid.autotune:158]:
[02:07:09][I][pid.autotune:159]:   control_parameters:
[02:07:09][I][pid.autotune:160]:     kp: 0.56172
[02:07:09][I][pid.autotune:161]:     ki: 0.00015
[02:07:09][I][pid.autotune:162]:     kd: 541.99701
[02:07:09][I][pid.autotune:163]:
[02:07:09][I][pid.autotune:164]:   Please copy these values into your YAML configuration! They will reset on the next reboot.
```

Observations:
- the ID in `PID Autotune - Iroda külső fűtés PID (iroda_kls_fts_pid):` is wrong. The ID of the component is `heat_01_pid`, not  `iroda_kls_fts_pid`
- as noted above a message stating the autotune has been started would be very useful. In my case I got the warning about heat/cool mode but if that would have not been the case, I would had no indication that it started something
- I think it would be nice to include the name or ID in _each_ line because when multiple autotunes run in parallel, their log messages are not queued into one, they might interleave if they wanna print out in the same time. I was thinking of something like adding the ID to the TAG:
```
[02:07:09][I][pid.autotune.heat_01_pid:119]: PID Autotune finished!
[02:07:09][I][pid.autotune.heat_01_pid:135]: PID Autotune - Iroda külső fűtés PID:
[02:07:09][I][pid.autotune.heat_01_pid:136]:   State: Succeeded!
[02:07:09][W][pid.autotune.heat_01_pid:144]:   Oscillation Frequency is not symmetrical. PID parameters may be inaccurate!
...
```

**Comment by gabormagyar** [comment]
Edit: This is incorrect, see https://github.com/esphome/esphome/pull/4136#issuecomment-1336554399

----

This might be better as a separate issue but  could you please also add a way to query the results after the autotune has completed?

I just completed two autotunes but weren't connected to the logs when the results were printed, so I don't know the results. Also I don't know if there's a way to get them back currently.

Could the logs maybe keep going and print the results periodically? The device is getting reinstalled after a successful autotune anyway.

**Comment by gabormagyar** [comment]
Also I agree that the initial message would be a good idea, maybe it could also include an explanation that initial temperature values are going to be `nan` until a fresh update is received from the sensor, as I described here:

https://github.com/esphome/issues/issues/3860#issue-1471090612

**Comment by nagyrobi** [comment]
> a way to query the results after the autotune has completed

It is possible by adding the Kp Ki Kd sensors to the config, before you do the autotune:
https://esphome.io/components/climate/pid.html#pid-sensor

**Comment by nagyrobi** [comment]
> Could the logs maybe keep going and print the results periodically?

Perhaps the results and the textual messages could be added to the purple config dump which is displayed when connecting to the logger via API. So any time one opens the log window immediately gets the results But that works only in DEBUG level mode... and that also rolls over pretty fast...

What I did before this patch was to set debug level to INFO, open a putty session to my HA as below, and keep it runnig until finished:

_In Home Assistant’s SSH addon, configure a username and a password, and disable Protection Mode (please assess the risks you take with that). Then, to access the logs form a device through an SSH client, log in, and you can use a command like:_
`docker exec -it addon_15ef4d2f_esphome esphome logs /config/esphome/pid_thermostats.yaml`
After finish no more roll-over, if only one autotune was running. The SSH window was waiting for me with the messages in it.

**Comment by mgabor3141** [comment]
I was pleasantly surprised on my second autotune that the results are in fact printed every time you connect after it has completed. The issue I described above is incorrect, what I experienced was most likely an interrupted autotune due to a brief power outage.

**Comment by gabormagyar** [comment]
I was pleasantly surprised on my second autotune that the results are in fact printed every time you connect after it has completed. The issue I described above is incorrect, what I experienced was most likely an interrupted autotune due to a brief power outage.

**Review by jesserockz — DISMISSED** [review]

**Comment by nagyrobi** [comment]
And yes! The results are printed in the logs after connecting later! Great, thank you @patrickcollins12 !
Although [my observations](https://github.com/esphome/esphome/pull/4136#issuecomment-1336117545) still stand, it's already much better!

Maybe a quick note, while autotune is still running, when connecting with a new logger client, print once something like `"Autotune is still running for PID %s:", name_.c_str()` so that we know that things are running fine behind the hood.

**Comment by patrickcollins12** [comment]
OK a few changes included to address the above requests. I only tested with one instance of a pid controller as I don't have a system with multiple PID controllers. @nagyrobi please test with multiple controllers.

Also I can confirm that it will now reprint the successful autotune results if you reconnect to the logs (so long as the ESP hasn't rebooted).

Here is an example of the logs.

```
[11:38:11][D][switch:013]: 'PID Climate Autotune' Turning ON.
[11:38:11][I][pid.climate.console_fan_thermostat:158]: Autotune has started. This can take a long time depending on the responsiveness of your system. Your system output will be altered to deliberately oscillate above and below the setpoint multiple times. Until your sensor provides a reading, the autotuner may display 'nan'
[11:38:11][W][pid.climate.console_fan_thermostat:166]: !!! For PID autotuner you need to set AUTO (also called heat/cool) mode! Console Fan Thermostat
[11:38:11][D][pid.autotune.console_fan_thermostat:184]: PID Autotune:
[11:38:11][D][pid.autotune.console_fan_thermostat:185]:   Autotune is still running!
[11:38:11][D][pid.autotune.console_fan_thermostat:186]:   Status: Trying to reach nan °C
[11:38:11][D][pid.autotune.console_fan_thermostat:187]:   Stats so far:
```

**Comment by nagyrobi** [comment]
Great! Thanks. Checkin today...

**Comment by nagyrobi** [comment]
Tested with multiple controllers, prints the internal ID to the log tags, and that accomplishes the request.

One very small suggestion: it would be very useful to reprint the `Autotune is still running!`  message when you reconnect to the logs while it's still running and there's no result yet. Some very slow cases can take long hours, reconnecting with the logs and seeing that would help confirming that no reboot or anything else interrupted it.

**Comment by patrickcollins12** [comment]
Great! @jesserockz can you please merge?

I'll consider this other request, not sure how to do that, but if so let's create another ticket for that.

**Comment by jesserockz** [comment]
Please fix the CI issues found.

Also, I dont like that the `TAG` has been overwritten with the id, the TAG should not be changed, instead you either need to prefix every log message, or put a "block start" / "block end" message before/after a group of logs

**Comment by patrickcollins12** [comment]
Hey @jesserockz ok i fixed this per your request not to alter the TAG. @nagyrobi please note that that formatting has changed a bit, but hopefully still works for you. It took me a while to please the almighty God of Clang.

File: esphome/components/pid/pid_autotuner.cpp [hunk 1]
```diff
@@ -221,7 +228,7 @@ float PIDAutotuner::RelayFunction::update(float error) {
   float output = state == RELAY_FUNCTION_POSITIVE ? output_positive : output_negative;
   if (change) {
     this->phase_count++;
-    ESP_LOGV(TAG, "Autotune: Turning output to %.1f%%", output * 100);
+    // ESP_LOGV(TAG, "Autotune: Turning output to %.1f%%", output * 100);
```
Comment (jesserockz) [inline esphome/components/pid/pid_autotuner.cpp]:
```suggestion
```

File: esphome/components/pid/pid_autotuner.cpp [hunk 2]
```diff
@@ -245,10 +252,10 @@ void PIDAutotuner::OscillationFrequencyDetector::update(uint32_t now, float erro
 
   if (had_crossing) {
     // Had crossing above hysteresis threshold, record
-    ESP_LOGV(TAG, "Autotune: Detected Zero-Cross at %u", now);
+    // ESP_LOGV(TAG, "Autotune: Detected Zero-Cross at %u", now);
```
Comment (jesserockz) [inline esphome/components/pid/pid_autotuner.cpp]:
```suggestion
```

File: esphome/components/pid/pid_autotuner.cpp [hunk 3]
```diff
@@ -245,10 +252,10 @@ void PIDAutotuner::OscillationFrequencyDetector::update(uint32_t now, float erro
...
     if (this->last_zerocross != 0) {
       uint32_t dt = now - this->last_zerocross;
-      ESP_LOGV(TAG, "  dt: %u", dt);
+      // ESP_LOGV(TAG, "  dt: %u", dt);
```
Comment (jesserockz) [inline esphome/components/pid/pid_autotuner.cpp]:
```suggestion
```

File: esphome/components/pid/pid_autotuner.cpp [hunk 4]
```diff
@@ -297,13 +304,13 @@ void PIDAutotuner::OscillationAmplitudeDetector::update(float error,
       // The positive error peak must have been in previous segment (180° shifted)
       // record phase_max
       this->phase_maxs.push_back(phase_max);
-      ESP_LOGV(TAG, "Autotune: Phase Max: %f", phase_max);
+      // ESP_LOGV(TAG, "Autotune: Phase Max: %f", phase_max);
```
Comment (jesserockz) [inline esphome/components/pid/pid_autotuner.cpp]:
```suggestion
```

File: esphome/components/pid/pid_autotuner.cpp [hunk 5]
```diff
@@ -297,13 +304,13 @@ void PIDAutotuner::OscillationAmplitudeDetector::update(float error,
...
     } else if (last_relay_state == RelayFunction::RELAY_FUNCTION_NEGATIVE) {
       // Transitioned from negative error to positive error.
       // The negative error peak must have been in previous segment (180° shifted)
       // record phase_min
       this->phase_mins.push_back(phase_min);
-      ESP_LOGV(TAG, "Autotune: Phase Min: %f", phase_min);
+      // ESP_LOGV(TAG, "Autotune: Phase Min: %f", phase_min);
```
Comment (jesserockz) [inline esphome/components/pid/pid_autotuner.cpp]:
```suggestion
```

**Review by jesserockz — CHANGES_REQUESTED** [review]

**Comment by patrickcollins12** [comment]
damn, sorry to deploy those two files... my bad, thanks for deleting.

didn't like those log entries? :)

**Comment by patrickcollins12** [comment]
ready @jesserockz

**Review by jesserockz — APPROVED** [review]

**Pull request merged**